File: packages/python-trace/trace.py
import logging
import os
import threading
import time
import uuid
import json
from datetime import datetime, timezone

logger = logging.getLogger("spolm.trace")


class Tracer:

    def __init__(self, api_key, agent_id, user_id="default", options=None, debug=False, spolm=None):
        if not api_key:
            raise ValueError("api_key is required")
        if not agent_id:
            raise ValueError("agent_id is required")

        self.API_KEY = api_key
        self.AGENT_ID = agent_id
        self.user_id = user_id
        self.options = options or {}
        self.debug = debug
        self.current_run = None
        self._spolm = spolm

        threading.Thread(target=self._validate_api_key, daemon=True).start()

    def _validate_api_key(self):
        try:
            from api.keys import check_api_key
            result = check_api_key(self.API_KEY)
            if not result.get("valid", False):
                logger.warning("Invalid Spolm API key — calls will fail")
        except Exception as e:
            logger.warning("Could not validate API key: %s", e)

    def start_run(self, user_task, metadata=None):
        if not user_task:
            raise ValueError("user_task is required to start a run")

        self.current_run = {
            "run_id": str(uuid.uuid4()),
            "user_id": self.user_id,
            "agent_id": self.AGENT_ID,
            "start_timestamp": datetime.now(timezone.utc).isoformat(),
            "user_task": user_task,
            "metadata": metadata or {},
            "steps": [],
            "final_output": None,
            "duration": 0,
        }

        return self.current_run["run_id"]

    def log_step(self, *, step_name, step_type, provider=None, options=None):
        if not step_name or not step_type:
            raise ValueError("step_name and step_type are required")

        def decorator(fn):
            tracer = self

            async def wrapper(*args, **kwargs):
                if not tracer.current_run:
                    raise RuntimeError("No active run. Call start_run first.")

                step_start = time.time()
                step_id = str(uuid.uuid4())
                input_data = args[0] if len(args) == 1 else list(args)

                tracer._step_meta = {}

                def record_tokens(usage):
                    tracer._step_meta["tokenUsage"] = usage

                tracer.record_tokens = record_tokens

                result = None
                error = None

                try:
                    result = await fn(*args, **kwargs)
                except Exception as err:
                    error = err

                step = {
                    "step_id": step_id,
                    "step_name": step_name,
                    "step_type": step_type,
                    "step_input": input_data,
                    "step_output": None if error else result,
                    "step_error": str(error) if error else None,
                    "step_end_time": datetime.now(timezone.utc).isoformat(),
                    "step_latency": int((time.time() - step_start) * 1000),
                    "tool_provider": provider,
                    "options": options or {},
                    "step_status": "failed" if error else "success",
                }

                if "tokenUsage" in tracer._step_meta:
                    step["tokens"] = tracer._step_meta["tokenUsage"]

                tracer.current_run["steps"].append(step)

                del tracer._step_meta
                del tracer.record_tokens

                if error:
                    raise error

                return result

            return wrapper

        return decorator

    def end_run(self, final_result, status="complete"):
        if not self.current_run:
            raise RuntimeError("No active run. Call start_run first.")

        end_timestamp = datetime.now(timezone.utc).isoformat()
        self.current_run["final_output"] = final_result
        self.current_run["status"] = status
        self.current_run["end_timestamp"] = end_timestamp

        start = datetime.fromisoformat(self.current_run["start_timestamp"])
        end = datetime.fromisoformat(end_timestamp)
        self.current_run["duration"] = int((end - start).total_seconds() * 1000)

        if self.debug:
            self._write_debug_file()

        self._post_log()

        if self._spolm:
            self._trigger_learning(final_result, status)

    def _trigger_learning(self, final_result, status):
        outcome = "success" if status == "complete" else "failure"
        run = self.current_run
        threading.Thread(
            target=self._spolm.record,
            kwargs={
                "task": run["user_task"],
                "outcome": outcome,
                "trajectory": run["steps"],
                "result": final_result,
            },
            daemon=True,
        ).start()

    def _post_log(self):
        try:
            from api.logs import post_log
            post_log(self.API_KEY, self.AGENT_ID, self.current_run)
        except Exception as e:
            logger.warning("Failed to post log: %s", e)

    def _write_debug_file(self):
        try:
            debug_dir = os.path.join(os.path.dirname(__file__), "..", "debug_logs")
            os.makedirs(debug_dir, exist_ok=True)
            ts = datetime.now(timezone.utc).isoformat().replace(":", "-").replace(".", "-")
            path = os.path.join(debug_dir, f"log_{ts}.json")
            with open(path, "w", encoding="utf-8") as f:
                json.dump(self.current_run, f, indent=2)
            logger.debug("Debug log written to %s", path)
        except Exception as e:
            logger.warning("Failed to write debug log: %s", e)

File: packages/js-trace/trace.js
const { v4: uuidv4 } = require("uuid");
const { checkAPIKey } = require("./api/keys.js");
const { postLog } = require("./api/logs.js");
const fs = require("fs");
const path = require("path");

class Tracer {
  /**
   * @param {string} apiKey - Spolm API key
   * @param {string} agentId - Agent identifier
   * @param {Object} options
   * @param {string} options.userId - User identifier for multi-tenancy (default: "default")
   * @param {boolean} options.debug - Write debug files to disk (default: false)
   * @param {Object} options.spolm - Spolm learning SDK instance for auto-learning on endRun()
   * @param {string} options.baseUrl - Override the API base URL (self-hosted only, default: "https://api.tryspolm.com")
   */
  constructor(apiKey, agentId, { userId = "default", debug = false, spolm = null, baseUrl = null } = {}) {
    if (!apiKey) throw new Error("apiKey is required");
    if (!agentId) throw new Error("agentId is required");

    this.apiKey = apiKey;
    this.agentId = agentId;
    this.userId = userId;
    this.debug = debug;
    this._spolm = spolm;
    this.currentRun = null;
    this.baseUrl = baseUrl || process.env.SPOLM_BASE_URL || "https://api.tryspolm.com";

    // fire-and-forget validation — never blocks constructor
    checkAPIKey(apiKey, this.baseUrl).then((result) => {
      if (!result.valid) {
        console.warn("[spolm] Invalid API key — calls will fail");
      }
    }).catch(() => {});
  }

  startRun(userTask, metadata = {}) {
    if (!userTask) throw new Error("userTask is required to start a run");

    this.currentRun = {
      run_id: uuidv4(),
      user_id: this.userId,
      agent_id: this.agentId,
      start_timestamp: new Date().toISOString(),
      user_task: userTask,
      metadata,
      steps: [],
      final_output: null,
      duration: 0,
    };

    return this.currentRun.run_id;
  }

  logStep({ stepName, stepType, provider = null, options = {} } = {}) {
    if (!stepName || !stepType) {
      throw new Error("stepName and stepType are required");
    }

    return (fn) => {
      const tracer = this;

      return async function (...args) {
        if (!tracer.currentRun) throw new Error("No active run. Call startRun first.");

        const stepStart = Date.now();
        const stepId = uuidv4();
        const input = args.length === 1 ? args[0] : args;

        tracer._stepMeta = {};
        tracer.recordTokens = (usage) => {
          tracer._stepMeta.tokenUsage = usage;
        };

        let result, error;
        try {
          result = await fn.apply(tracer, args);
        } catch (err) {
          error = err;
        }

        const step = {
          step_id: stepId,
          step_name: stepName,
          step_type: stepType,
          step_input: input,
          step_output: error ? null : result,
          step_error: error ? error.message : null,
          step_end_time: new Date().toISOString(),
          step_latency: Date.now() - stepStart,
          tool_provider: provider,
          options,
          step_status: error ? "failed" : "success",
          ...(tracer._stepMeta.tokenUsage && { tokens: tracer._stepMeta.tokenUsage }),
        };

        tracer.currentRun.steps.push(step);

        delete tracer._stepMeta;
        delete tracer.recordTokens;

        if (error) throw error;
        return result;
      };
    };
  }

  endRun(finalResult, status = "complete") {
    if (!this.currentRun) throw new Error("No active run. Call startRun first.");

    const endTimestamp = new Date().toISOString();
    this.currentRun.final_output = finalResult;
    this.currentRun.status = status;
    this.currentRun.end_timestamp = endTimestamp;
    this.currentRun.duration = new Date(endTimestamp) - new Date(this.currentRun.start_timestamp);

    if (this.debug) {
      this._writeDebugFile();
    }

    this._postLog();

    if (this._spolm) {
      this._triggerLearning(finalResult, status);
    }
  }

  _triggerLearning(finalResult, status) {
    const outcome = status === "complete" ? "success" : "failure";
    const run = this.currentRun;
    // fire and forget — never blocks endRun
    Promise.resolve().then(() =>
      this._spolm.record({
        task: run.user_task,
        outcome,
        trajectory: run.steps,
        result: finalResult,
      })
    ).catch(() => {});
  }

  _postLog() {
    postLog(this.apiKey, this.agentId, this.currentRun, this.baseUrl).catch(() => {});
  }

  _writeDebugFile() {
    try {
      const debugDir = path.resolve(__dirname, "../debug-logs");
      if (!fs.existsSync(debugDir)) fs.mkdirSync(debugDir, { recursive: true });
      const ts = new Date().toISOString().replace(/[:.]/g, "-");
      fs.writeFileSync(
        path.join(debugDir, `log_${ts}.json`),
        JSON.stringify(this.currentRun, null, 2),
        "utf8"
      );
    } catch (e) {
      console.warn("[spolm] Failed to write debug log:", e.message);
    }
  }
}

module.exports = Tracer;
